### PR TITLE
Basic support for multi-folder deployment creation, selection and deployment

### LINKED ISSUE #1852: List recursive options for deployments and entrypoints
With the new recursive APIs we can show all of the options, even ones in nested folders.

The Deployment selection should use the `?recursive=true` query param to show all Deployments.

When creating a new Deployment, the entrypoints listed should use `?recursive=true` to show the user all entrypoints. (This will be adjusted in the future, but is the simplest solution right now).

---

Part of the #1849 epic

### PR DESCRIPTION
**Pull request by sagerb**
## Intent

Resolves #1852 

## Type of Change

- - [ ] Bug Fix 
- - [x] New Feature 
- - [ ] Breaking Change 
- - [ ] Documentation 
- - [x] Refactor 
- - [ ] Tooling 

## Approach

The changes needed to implement this proved to be pretty much as expected.

1. Needed to take into account projectDir when selecting or matching configurations and deployments
2. Needed to update API calls to be recursive
3. Needed to update new deployment process to be able to create unique names per the project directory
4. File watchers needed to watch recursively
5. Needed to pass in projectDir when calling deploy api

## Directions for Reviewers

Unfortunately, there were a number of changes required to maintain functionality once changes were started. It would be best to walk through the individual commits, to understand what is being done.

[recursive inspection, create deployment and configuration within dir …](https://github.com/posit-dev/publisher/pull/1943/commits/3e6d08028476d94f101f29d78ce6a45460713be1)

[Update watcher patterns to be recursive](https://github.com/posit-dev/publisher/pull/1943/commits/bc4cd805d262fa2ca685f9a1158530a79a29691d)

[Recursive API calls, integrate projectDir and switch to deploymentSel…](https://github.com/posit-dev/publisher/pull/1943/commits/263d27f2458c3fc881530fc1f4c6dd9c3a0d8392)

[Deploy with projectDir parameter (and update deployment tooltip)](https://github.com/posit-dev/publisher/pull/1943/commits/fb7b6adcc47ef4a1b5b83084704cae2dfe1d77d1)

[Fix codicons path issue (underscore was lost during search/replace)](https://github.com/posit-dev/publisher/pull/1943/commits/e9781c4a2e55611c41eb939abbe992d2dcd344ce)

## Checklist

- [ ] I have updated [CHANGELOG.md](../CHANGELOG.md) to cover notable changes.

### CODE REVIEW

File: extensions/vscode/src/constants.ts [hunk 1]
```diff
@@ -1,11 +1,11 @@
 // Copyright (C) 2024 by Posit Software, PBC.
 
-export const POSIT_FOLDER = ".posit";
-export const PUBLISH_FOLDER = ".posit/publish";
-export const PUBLISH_DEPLOYMENTS_FOLDER = ".posit/publish/deployments";
+export const POSIT_FOLDER = "**/.posit";
```
Comment (sagerb) [inline extensions/vscode/src/constants.ts:3]: All of our file watcher patterns need now to glob downward into subdirectories.

File: extensions/vscode/src/multiStepInputs/newDeployment.ts [hunk 2]
```diff
@@ -267,7 +267,7 @@ export async function newDeployment(
 
   let entryPointListItems: QuickPickItemWithIndex[] = [];
   let inspectionResults: ConfigurationInspectionResult[] = [];
-  let contentRecordNames: string[] = [];
+  let contentRecordNames = new Map<string, string[]>();
```
Comment (sagerb) [inline extensions/vscode/src/multiStepInputs/newDeployment.ts:270]: switch to a Map here so we can have easy filtering of configurations which are applicable to a projectDir where the entrypoints and deployment is being created.

File: extensions/vscode/src/multiStepInputs/newDeployment.ts [hunk 3]
```diff
@@ -367,6 +369,7 @@ export async function newDeployment(
               iconPath: new ThemeIcon("file"),
               label: config.entrypoint,
               description: `(${contentTypeStrings[config.type]})`,
+              detail: result.projectDir,
```
Comment (sagerb) [inline extensions/vscode/src/multiStepInputs/newDeployment.ts:372]: We needed to indicate the location of the entrypoints, so now I'm using the second line to hold the workingDir

File: extensions/vscode/src/multiStepInputs/newDeployment.ts [hunk 4]
```diff
@@ -392,12 +395,17 @@ export async function newDeployment(
...
-        (contentRecord) => contentRecord.deploymentName,
-      );
+      contentRecordList.forEach((contentRecord) => {
+        let existingList = contentRecordNames.get(contentRecord.projectDir);
+        if (existingList === undefined) {
+          existingList = [];
+        }
+        existingList.push(contentRecord.deploymentName);
+        contentRecordNames.set(contentRecord.projectDir, existingList);
+      });
```
Comment (sagerb) [inline extensions/vscode/src/multiStepInputs/newDeployment.ts:408]: Populate the map to be keyed by a projectDir and then contain an array of entrypoints.

File: extensions/vscode/src/multiStepInputs/newDeployment.ts [hunk 5]
```diff
@@ -896,20 +904,22 @@ export async function newDeployment(
...
+  const selectedInspectionResult =
+    inspectionResults[state.data.entryPoint.index];
+  if (!selectedInspectionResult) {
+    window.showErrorMessage(
+      `Unable to proceed creating configuration. Error retrieving config for ${state.data.entryPoint.label}, index = ${state.data.entryPoint.index}`,
+    );
+    return;
+  }
+  selectedInspectionResult.configuration.title = state.data.title;
+
```
Comment (sagerb) [inline extensions/vscode/src/multiStepInputs/newDeployment.ts:916]: same logic, just moved around a bit.

File: extensions/vscode/src/multiStepInputs/newDeployment.ts [hunk 6]
```diff
@@ -948,10 +958,17 @@ export async function newDeployment(
 
   // Create the PrecontentRecord File
   try {
+    let existingNames = contentRecordNames.get(
+      selectedInspectionResult.projectDir,
+    );
+    if (!existingNames) {
+      existingNames = [];
+    }
```
Comment (sagerb) [inline extensions/vscode/src/multiStepInputs/newDeployment.ts:966]: easy use of our Map

File: extensions/vscode/src/types/messages/hostToWebviewMessages.ts [hunk 7]
```diff
@@ -71,15 +72,15 @@ export type RefreshContentRecordDataMsg = AnyHostToWebviewMessage<
   HostToWebviewMessageType.REFRESH_CONTENTRECORD_DATA,
   {
     contentRecords: (ContentRecord | PreContentRecord)[];
-    selectedContentRecordName?: string | null;
+    deploymentSelected: DeploymentSelector | null;
```
Comment (sagerb) [inline extensions/vscode/src/types/messages/hostToWebviewMessages.ts:75]: switch to a DeploymentSelector object, which allows us to select the exact combination of deployment and config. It is unique across all locations within a set of directories, since it contains the actual deployment path on disk.

File: extensions/vscode/src/types/messages/webviewToHostMessages.ts [hunk 8]
```diff
@@ -89,6 +89,7 @@ export type DeployMsg = AnyWebviewToHostMessage<
     deploymentName: string;
     credentialName: string;
     configurationName: string;
+    projectDir: string;
```
Comment (sagerb) [inline extensions/vscode/src/types/messages/webviewToHostMessages.ts:92]: New parameter needed to deploy!

File: extensions/vscode/src/types/shared.ts [hunk 9]
```diff
@@ -7,16 +7,14 @@ import {
   PreContentRecord,
 } from "../api";
 
-export type HomeViewState = {
-  deploymentName?: string;
-  configurationName?: string;
-};
-
-export type DeploymentNames = {
+export type DeploymentSelector = {
```
Comment (sagerb) [inline extensions/vscode/src/types/shared.ts:10]: Replaces use of DeploymentNames (and HomeViewState)

File: extensions/vscode/src/views/homeView.ts [hunk 10]
```diff
@@ -354,7 +357,7 @@ export class HomeViewProvider implements WebviewViewProvider, Disposable {
   private async refreshConfigurationData() {
     try {
       const api = await useApi();
-      const response = await api.configurations.getAll();
+      const response = await api.configurations.getAll({ recursive: true });
```
Comment (sagerb) [inline extensions/vscode/src/views/homeView.ts:360]: recursively refreshing everytime.. We probably can figure out a more targeted approach, but lets do that in another PR as an optimization.

File: extensions/vscode/src/multiStepInputs/newDeployment.ts (hunk 2 above)
Comment (dotNomad) [reply extensions/vscode/src/multiStepInputs/newDeployment.ts:270]: I wonder if it would be helpful in other inputs to have a util class for names to add them, use the `projectDir` to gather them and all that. Something to consider if we repeat the logic elsewhere.

File: extensions/vscode/src/types/messages/hostToWebviewMessages.ts [hunk 11]
```diff
@@ -71,15 +72,15 @@ export type RefreshContentRecordDataMsg = AnyHostToWebviewMessage<
...
   }
 >;
 export type RefreshConfigDataMsg = AnyHostToWebviewMessage<
   HostToWebviewMessageType.REFRESH_CONFIG_DATA,
   {
     configurations: Configuration[];
     configurationsInError: ConfigurationError[];
-    selectedConfigurationName?: string | null;
+    deploymentSelected: DeploymentSelector | null;
```
Comment (dotNomad) [inline extensions/vscode/src/types/messages/hostToWebviewMessages.ts:83]: Is this no longer optional?

File: extensions/vscode/webviews/homeView/src/stores/home.ts [hunk 12]
```diff
@@ -85,12 +89,18 @@ export const useHomeStore = defineStore("home", () => {
    * @param name the name of the new configuration to select
    * @returns true if the selected contentRecord was the same, false if not
    */
-  function updateSelectedConfigurationByName(name: string) {
+  function updateSelectedConfigurationByName(
+    name: string,
+    projectDir?: string,
```
Comment (dotNomad) [inline extensions/vscode/webviews/homeView/src/stores/home.ts:94]: I'm wondering if this should be optional to avoid us changing a selected configuration incorrectly.

File: extensions/vscode/webviews/homeView/src/stores/home.ts [hunk 13]
```diff
@@ -113,15 +123,23 @@ export const useHomeStore = defineStore("home", () => {
...
-      kind: WebviewToHostMessageType.SAVE_SELECTION_STATE,
-      content: {
-        state: {
-          deploymentName: selectedContentRecord.value?.saveName,
-          configurationName: selectedConfiguration.value?.configurationName,
+    let state: HomeViewState = null;
+    // skip saving the selection if we have not yet
+    // selected a content record. This is probably an init
+    // sequence.
+    if (selectedContentRecord.value) {
```
Comment (dotNomad) [inline extensions/vscode/webviews/homeView/src/stores/home.ts:130]: Bit confused by this change. Wouldn't we want this to update if the deployment selected was deleted for example?

File: extensions/vscode/src/views/homeView.ts [hunk 14]
```diff
@@ -430,36 +433,49 @@ export class HomeViewProvider implements WebviewViewProvider, Disposable {
     });
   }
 
-  private getSelectionState(): HomeViewState {
-    const state = this.context.workspaceState.get<HomeViewState>(
+  private getSelectionState(): DeploymentSelector | null {
```
Comment (dotNomad) [inline extensions/vscode/src/views/homeView.ts:436]: Shouldn't this still be `HomeViewState` since that type is `DeploymentSelector | null` anyway?

File: extensions/vscode/src/views/homeView.ts [hunk 15]
```diff
@@ -675,20 +701,16 @@ export class HomeViewProvider implements WebviewViewProvider, Disposable {
   }
 
   private async propogateDeploymentSelection(
-    configurationName?: string,
-    deploymentName?: string,
+    documentSelector: DeploymentSelector | null,
```
Comment (mmarchetti) [inline extensions/vscode/src/views/homeView.ts:704]: nit: this feels like it should be named `deploymentSelector`

**Review by dotNomad — APPROVED** [review]
Had a couple concerns, but overall this is looking great - we can address as follow-ups if need be.

**Review by mmarchetti — APPROVED** [review]

File: extensions/vscode/webviews/homeView/src/stores/home.ts (hunk 13 above)
Comment (sagerb) [reply extensions/vscode/webviews/homeView/src/stores/home.ts:130]: If the config selected is always part of the deployment, if that does not exist, then we wouldn't want to update the config selected (since we have no deployment record).

File: extensions/vscode/webviews/homeView/src/stores/home.ts (hunk 12 above)
Comment (sagerb) [reply extensions/vscode/webviews/homeView/src/stores/home.ts:94]: should be or shouldn't be? I'm thinking shouldn't be, for the reason you mention?

File: extensions/vscode/src/types/messages/hostToWebviewMessages.ts (hunk 11 above)
Comment (sagerb) [reply extensions/vscode/src/types/messages/hostToWebviewMessages.ts:83]: no longer needed. It is always updated within the deployment selected, as we are writing out the deployment record on update of the selected configuration.

File: extensions/vscode/webviews/homeView/src/stores/home.ts (hunk 12 above)
Comment (dotNomad) [reply extensions/vscode/webviews/homeView/src/stores/home.ts:94]: Yes sorry, _shouldn't_ be optional

File: extensions/vscode/webviews/homeView/src/stores/home.ts (hunk 13 above)
Comment (dotNomad) [reply extensions/vscode/webviews/homeView/src/stores/home.ts:130]: Gotcha, that makes more sense to me 👍
